### PR TITLE
Structural changes to apply tags search

### PR DESCRIPTION
**Pull request by stat-kwon**
### Category
- [ ] New feature
- [ ] Bug fix
- [x] Improvement
- [x] Refactor
- [ ] etc

### Description
Structural changes to apply tags search
* remove the CloudServiceTag
* Apply namespaces to some fields
  * tags
  * metadata
  * collection_info 

### Known issue
* #30

### CODE REVIEW

File: src/spaceone/inventory/service/cloud_service_service.py [hunk 1]
```diff
@@ -181,30 +176,42 @@ def update(self, params):
         old_cloud_svc_data = dict(cloud_svc_vo.to_dict())
 
         if 'tags' in params:
-            new_tags = self._create_tag(params, tag_type)
+            old_tags = old_cloud_svc_data.get('tags', {})
+            new_tags = self._create_tags(params['tags'], provider)
```
Comment (whdalsrnt) [inline src/spaceone/inventory/service/cloud_service_service.py:180]: The two methods seem to perform the same action, so please combine them into one.
And the function name looks better `convert` than `create`.

File: src/spaceone/inventory/service/cloud_service_service.py [hunk 2]
```diff
@@ -181,30 +176,42 @@ def update(self, params):
...
+                params['tags'] = old_tags
+                params['tag_keys'] = old_tag_keys
+            else:
+                del params['tags']
 
-            new_tags_dict: dict = self._change_tags_to_dict(new_tags)
-            old_tags_dict: dict = self._change_tags_to_dict(old_cloud_svc_data['tags'], tag_type=tag_type)
+        if 'metadata' in params:
+            old_metadata = old_cloud_svc_data.get('metadata', {})
+            new_metadata = self._change_metadata_path(params['metadata'], provider)
```
Comment (whdalsrnt) [inline src/spaceone/inventory/service/cloud_service_service.py:194]: Again, the function name `convert` looks better than `change`.

File: src/spaceone/inventory/service/cloud_service_service.py [hunk 3]
```diff
@@ -181,30 +176,42 @@ def update(self, params):
         old_cloud_svc_data = dict(cloud_svc_vo.to_dict())
 
         if 'tags' in params:
-            new_tags = self._create_tag(params, tag_type)
+            old_tags = old_cloud_svc_data.get('tags', {})
+            new_tags = self._create_tags(params['tags'], provider)
+            old_tag_keys = old_cloud_svc_data.get('tag_keys', {})
+            new_tag_keys = self._create_tag_keys(params['tags'], provider)
+
+            if new_tags != old_tags:
```
Comment (whdalsrnt) [inline src/spaceone/inventory/service/cloud_service_service.py:184]: Comparison of two tags should be limited to the input provider tags, not all tags.
`if new_tags[provider] != old_tags.get(provider, {}):`

File: src/spaceone/inventory/service/cloud_service_service.py [hunk 4]
```diff
@@ -181,30 +176,42 @@ def update(self, params):
...
+                old_tags.update(new_tags)
+                old_tag_keys.update(new_tag_keys)
+                params['tags'] = old_tags
+                params['tag_keys'] = old_tag_keys
```
Comment (whdalsrnt) [inline src/spaceone/inventory/service/cloud_service_service.py:188]: Declare a set to remove duplicate keys.
`params['tag_keys'] = list(set(old_tags_keys))`

File: src/spaceone/inventory/service/cloud_service_service.py [hunk 5]
```diff
@@ -181,30 +176,42 @@ def update(self, params):
...
-            elif 'tags' in params:
-                del params['tags']
+            if new_metadata != old_metadata:
+                old_metadata.update(new_metadata)
+                params['metadata'] = old_metadata
+            else:
+                del params['metadata']
 
-        params['collection_info'] = self._get_collection_info(old_cloud_svc_data.get('collection_info', {}))
+        if 'collection_info' in params:
```
Comment (whdalsrnt) [inline src/spaceone/inventory/service/cloud_service_service.py:202]: collection_info is not declared in params, so change the logic.

File: src/spaceone/inventory/service/cloud_service_service.py [hunk 6]
```diff
@@ -454,101 +462,98 @@ def _make_cloud_service_type_key(resource_data):
...
         return f'{resource_data["domain_id"]}.{provider}.{resource_data["region_code"]}'
 
-    def _change_metadata_path(self, metadata):
+    def _change_metadata_path(self, metadata, provider=None):
+        if provider is None:
+            provider = 'custom'
+
         plugin_id = self.transaction.get_meta('plugin_id')
 
         if plugin_id:
```
Comment (whdalsrnt) [inline src/spaceone/inventory/service/cloud_service_service.py:471]: change condition
`plugin_id` -> `provider`

**Review by whdalsrnt — CHANGES_REQUESTED** [review]

File: src/spaceone/inventory/info/cloud_service_info.py [hunk 7]
```diff
@@ -32,7 +32,8 @@ def CloudServiceInfo(cloud_svc_vo: CloudService, minimal=False):
             'data': change_struct_type(cloud_svc_vo.data),
             'metadata': change_struct_type(cloud_svc_vo.metadata),
             'tags': change_struct_type(_tags_to_dict(cloud_svc_vo.tags)),
-            'collection_info': CollectionInfo(cloud_svc_vo.collection_info.to_dict()),
+            'tag_keys': change_struct_type(cloud_svc_vo.tag_keys),
+            'collection_info': CollectionInfo(cloud_svc_vo.collection_info),
```
[the remark this answers is not in the document]
Comment (stat-kwon) [reply src/spaceone/inventory/info/cloud_service_info.py:53]: have reflected thank you.

File: src/spaceone/inventory/info/cloud_service_info.py [hunk 8]
```diff
@@ -46,8 +47,11 @@ def CloudServicesInfo(cloud_svc_vos, total_count, **kwargs):
         map(functools.partial(CloudServiceInfo, **kwargs), cloud_svc_vos)), total_count=total_count)
 
 
-def _tags_to_dict(tags: list) -> dict:
-    dict_value = {}
-    for tag in tags:
-        dict_value[tag['key']] = tag['value']
-    return dict_value
+def _tags_to_dict(tags) -> dict:
```
[the remark this answers is not in the document]
Comment (stat-kwon) [reply src/spaceone/inventory/info/cloud_service_info.py:67]: I changed from `_tag_to_dict` to `_change_tags_without_hash` and fixed the logic. 
The meaning of function is clearer than before. thank you.

File: src/spaceone/inventory/service/cloud_service_service.py [hunk 9]
```diff
@@ -70,15 +69,16 @@ def create(self, params):
...
-        params['provider'] = params.get('provider', self.transaction.get_meta('secret.provider'))
+        params['provider'] = provider
 
         if params['provider'] is None:
             raise ERROR_REQUIRED_PARAMETER(key='provider')
 
         if 'tags' in params:
-            tag_type = self._get_tag_type_from_meta()
-            params['tags'] = self._create_tag(params, tag_type)
+            params['tag_keys'] = self._create_tag_keys(params['tags'], provider)
```
[the remark this answers is not in the document]
Comment (stat-kwon) [reply src/spaceone/inventory/service/cloud_service_service.py:80]: Combined `_create_tags` and `_create_tag_keys` methods and changed to _convert_tag_type.
It has a clearer meaning than before. thank you

File: src/spaceone/inventory/service/cloud_service_service.py (hunk 3 above)
Comment (stat-kwon) [reply src/spaceone/inventory/service/cloud_service_service.py:184]: have reflected thank you.

File: src/spaceone/inventory/service/cloud_service_service.py (hunk 1 above)
Comment (stat-kwon) [reply src/spaceone/inventory/service/cloud_service_service.py:180]: have reflected thank you.

File: src/spaceone/inventory/service/cloud_service_service.py (hunk 2 above)
Comment (stat-kwon) [reply src/spaceone/inventory/service/cloud_service_service.py:194]: I changed from `_change_metadata_path` to `_convert_metadata` and fixed the logic.
The meaning of function is clearer than before. thank you.

File: src/spaceone/inventory/service/cloud_service_service.py (hunk 6 above)
Comment (stat-kwon) [reply src/spaceone/inventory/service/cloud_service_service.py:471]: have reflected thank you.

File: src/spaceone/inventory/info/cloud_service_info.py [hunk 10]
```diff
@@ -46,8 +64,12 @@ def CloudServicesInfo(cloud_svc_vos, total_count, **kwargs):
...
 
 
-def _tags_to_dict(tags: list) -> dict:
-    dict_value = {}
-    for tag in tags:
-        dict_value[tag['key']] = tag['value']
-    return dict_value
+def _change_tags_without_hash(tags) -> dict:
+    changed_tags = {}
+    for provider, tag in tags.items():
```
Comment (whdalsrnt) [inline src/spaceone/inventory/info/cloud_service_info.py:69]: change a syntax
```python
for provider, hashed_tag in tags.items():
        for hash_key, tag in hashed_tag.items():
            if not changed_tags.get(provider):
                changed_tags[provider] = {}
            changed_tags[provider][tag['key']] = tag['value']
```

File: src/spaceone/inventory/service/cloud_service_service.py [hunk 11]
```diff
@@ -70,15 +69,15 @@ def create(self, params):
         domain_id = params['domain_id']
         project_id = params.get('project_id')
         secret_project_id = self.transaction.get_meta('secret.project_id')
+        provider = params.get('provider', self.transaction.get_meta('secret.provider'))
```
Comment (whdalsrnt) [inline src/spaceone/inventory/service/cloud_service_service.py:72]: Move the logic to get the provider into a function.

File: src/spaceone/inventory/service/cloud_service_service.py [hunk 12]
```diff
@@ -181,30 +175,35 @@ def update(self, params):
...
                 del params['tags']
 
-        params['collection_info'] = self._get_collection_info(old_cloud_svc_data.get('collection_info', {}))
-
         if 'metadata' in params:
-            params['metadata'] = self._change_metadata_path(params['metadata'])
+            old_metadata = old_cloud_svc_data.get('metadata', {})
+            new_metadata = self._convert_metadata(params['metadata'], provider)
+
+            if new_metadata != old_metadata:
```
Comment (whdalsrnt) [inline src/spaceone/inventory/service/cloud_service_service.py:194]: change condition

`if self._is_different_data(new_xxx, old_xxx, provider):`

File: src/spaceone/inventory/service/cloud_service_service.py [hunk 13]
```diff
@@ -312,8 +311,9 @@ def list(self, params):
         query = params.get('query', {})
         query = self._append_resource_group_filter(query, params['domain_id'])
         query = self._change_project_group_filter(query, params['domain_id'])
-        query = self._change_tags_filter(query, params['domain_id'])
+        query = self._change_tags_filter(query)
```
Comment (whdalsrnt) [inline src/spaceone/inventory/service/cloud_service_service.py:314]: change function name
`_change_tags_filter` -> `_change_filter_tags`

File: src/spaceone/inventory/service/cloud_service_service.py [hunk 14]
```diff
@@ -454,101 +454,85 @@ def _make_cloud_service_type_key(resource_data):
     def _make_region_key(resource_data, provider):
         return f'{resource_data["domain_id"]}.{provider}.{resource_data["region_code"]}'
 
-    def _change_metadata_path(self, metadata):
-        plugin_id = self.transaction.get_meta('plugin_id')
-
-        if plugin_id:
+    @staticmethod
+    def _convert_metadata(metadata, provider):
```
Comment (whdalsrnt) [inline src/spaceone/inventory/service/cloud_service_service.py:458]: The logic to insert custom into the provider is implemented in the previous step.

File: src/spaceone/inventory/service/cloud_service_service.py [hunk 15]
```diff
@@ -454,101 +454,85 @@ def _make_cloud_service_type_key(resource_data):
...
+                'custom': copy.deepcopy(metadata)
             }
 
         return metadata
 
-    def _get_collection_info(self, collection_info=None):
-        if collection_info is None:
-            collection_info = {}
+    def _get_collection_info(self, provider, collections: CollectionInfo = None):
+        if collections is None:
```
Comment (whdalsrnt) [inline src/spaceone/inventory/service/cloud_service_service.py:471]: change a syntax

`collections = collections or []`

File: src/spaceone/inventory/service/cloud_service_service.py [hunk 16]
```diff
@@ -454,101 +454,85 @@ def _make_cloud_service_type_key(resource_data):
...
+            'provider': provider,
+            'collector_id': collector_id,
+            'secret_id': secret_id,
+            'service_account_id': service_account_id,
+            'last_collected_at': datetime.utcnow()
+        }
 
-        if collector_id:
-            all_collectors.append(collector_id)
+        collections.append(new_collection)
```
Comment (whdalsrnt) [inline src/spaceone/inventory/service/cloud_service_service.py:486]: Override the same provider data if there is an old CollectionInfo.

```
merged_collections = []
if collections:
  for collection in collections:
    if collection['provider'] == provider:
      merged_collections.append(new_collection)
    else:
      merged_collections.append(collection)
else:
   merged_collections.append(new_collection)
```

File: src/spaceone/inventory/service/cloud_service_service.py [hunk 17]
```diff
@@ -454,101 +454,85 @@ def _make_cloud_service_type_key(resource_data):
...
+        return collections
 
-        if service_account_id:
-            all_service_accounts.append(service_account_id)
+    @staticmethod
+    def _merge_collection_info(old_collection_info, new_collection_info):
```
Comment (whdalsrnt) [inline src/spaceone/inventory/service/cloud_service_service.py:490]: Currently this function seems to be deprecated.

File: src/spaceone/inventory/service/cloud_service_service.py [hunk 18]
```diff
@@ -454,101 +454,85 @@ def _make_cloud_service_type_key(resource_data):
...
-    def _change_tags_to_dict(tags: list, tag_type=None) -> dict:
-        tags_dict = {}
-        for tag in tags:
-            if tag_type and tag_type == tag['type']:
-                tags_dict[tag['key']] = tag['value']
-            else:
-                tags_dict[tag['key']] = tag['value']
+        tags = {provider: {}}
+        for key, value in dot_tags.items():
+            tags[provider].update({utils.string_to_hash(key): {'key': key, 'value': value}})
```
Comment (whdalsrnt) [inline src/spaceone/inventory/service/cloud_service_service.py:523]: change a syntax

```
hashed_key = utils.string_to_hash(key)
tags[provider][hashed_key] = {'key': key, 'value': value}
```

**Review by whdalsrnt — CHANGES_REQUESTED** [review]

File: src/spaceone/inventory/info/cloud_service_info.py (hunk 10 above)
Comment (stat-kwon) [reply src/spaceone/inventory/info/cloud_service_info.py:69]: have reflected thank you.

File: src/spaceone/inventory/service/cloud_service_service.py (hunk 11 above)
Comment (stat-kwon) [reply src/spaceone/inventory/service/cloud_service_service.py:72]: have reflected thank you.

File: src/spaceone/inventory/service/cloud_service_service.py (hunk 12 above)
Comment (stat-kwon) [reply src/spaceone/inventory/service/cloud_service_service.py:194]: have reflected thank you.

File: src/spaceone/inventory/service/cloud_service_service.py (hunk 13 above)
Comment (stat-kwon) [reply src/spaceone/inventory/service/cloud_service_service.py:314]: have reflected thank you.

File: src/spaceone/inventory/service/cloud_service_service.py (hunk 14 above)
Comment (stat-kwon) [reply src/spaceone/inventory/service/cloud_service_service.py:458]: have reflected thank you.

File: src/spaceone/inventory/service/cloud_service_service.py (hunk 15 above)
Comment (stat-kwon) [reply src/spaceone/inventory/service/cloud_service_service.py:471]: have reflected thank you.

File: src/spaceone/inventory/service/cloud_service_service.py (hunk 16 above)
Comment (stat-kwon) [reply src/spaceone/inventory/service/cloud_service_service.py:486]: have reflected thank you.

File: src/spaceone/inventory/service/cloud_service_service.py (hunk 17 above)
Comment (stat-kwon) [reply src/spaceone/inventory/service/cloud_service_service.py:490]: have reflected thank you.

File: src/spaceone/inventory/service/cloud_service_service.py (hunk 18 above)
Comment (stat-kwon) [reply src/spaceone/inventory/service/cloud_service_service.py:523]: have reflected thank you.

File: src/spaceone/inventory/service/cloud_service_service.py [hunk 19]
```diff
@@ -557,59 +541,45 @@ def _change_tags_filter(self, query, domain_id):
             operator = condition.get('o', condition.get('operator'))
 
             if key.startswith('tags.'):
-                tag_key = key.replace('tags.', '')
+                prefix, provider, key = key.split('.', 2)
```
[the remark this answers is not in the document]
Comment (stat-kwon) [reply src/spaceone/inventory/service/cloud_service_service.py]: have reflected thank you.

**Review by whdalsrnt — APPROVED** [review]
LGTM!!

**Pull request merged**